File: cs50/test/js/cash.js
// Airbnb JavaScript style guide: https://github.com/airbnb/javascript/

/**
 * Prints the minimum US coins (up to quarters) needed to make a given
 * amount of change owed in cents.
 */


// Import the readline-sync module to interact with user via console.
const readlineSync = require('readline-sync');

// Main function.
const main = function mainFunction () {
  console.log();

  // Get change owed.
  let userChange = getChange('Enter change (in dollars): ');

  // Count coins necessary for user's change.
  let coins = countCoins(userChange);

  // Print number of coins (as an integer and a newline).
  console.log(coins);

  console.log();
}

// Return user-input positive int amount of change in dollars.
const getChange = function getChangeFunction (prompt) {
  let change;

  // Continually reprompt if change input is not valid.
  while (true) {
    // Get user input.
    change = readlineSync.question(prompt);

    // Check if input is empty.
    if (change === "") {
      console.log('error: no change entered\n');
      continue;
    }

    // Convert change to a number.
    change = Number(change)

    // Check if input is in dollars.2d format.
    if (!Number.isInteger(change * 100)) {
      console.log('error: invalid dollars and cents\n');
      continue;

    // Check if input is negative.
    } else if (change < 0) {
      console.log('change must be positive');
      continue;
    }

    return change;
  }
}

// Count the minimum number of coins equal to argument amount.
const countCoins = function countCoinsFunction (dollars) {
  // List of coin values from quarters down.
  const coinValues = [25, 10, 5, 1];

  // Convert float into integer coin value for calculations.
  let cents = parseInt(dollars * 100);

  // Initialize counter variable.
  let count = 0;

  // Iterate through list of coins.
  coinValues.forEach(function(coin) {
    while (true) {
      // Switch to smaller coin if necessary.
      if (cents < coin) {
          break;
      }

      // Subtract value of each coin; count the coin.
      cents -= coin;
      count += 1;
    }
  })

  // Return count when done subtracting coins.
  return count;
}

main();

File: cs50/week1/cash/cash.c
// Prints the minimum US coins (up to quarters) needed to make a given amount of change owed in cents

#include <stdio.h>
#include <cs50.h>

// Continually prompts user for positive int amount of change in cents
int get_change(void);

// Lowers input coin value to next highest coin
int lower_coin(int coin);

int main(void)
{
    // Get change owed
    int change = get_change();

    // Count number of coins with n
    int n = 0;

    // Subtract largest possible coin and increment n by 1, starting with quarter
    int coin = 25;
    while (change > 0)
    {
        // Compare change value to highest coin value and subtract/increment if possible
        if (change >= coin)
        {
            change -= coin;
            n++;
        }
        else
        {
            // Lower coin value to next highest and loop until change is 0
            coin = lower_coin(coin);
        }
    }

    // Print final number of coins
    printf("%i\n", n);
}

// Continually prompts user for positive int amount of change in cents
int get_change(void)
{
    int change;
    do
    {
        change = get_int("Change owed: ");
    }
    while (change < 0);
    return change;
}

// Lowers coin value to next highest coin
int lower_coin(int coin)
{
    int lower = 0;
    if (coin == 25)
    {
        lower = 10;
    }
    else if (coin == 10)
    {
        lower = 5;
    }
    else if (coin == 5)
    {
        lower = 1;
    }
    else
    {
        printf("Error: invalid coin value\n");
    }
    return lower;
}